kroxylicious-api/src/main/java/io/kroxylicious/proxy/filter/FilterDispatchExecutor.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter;
+
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * An Executor backed by the <b>Filter Dispatch Thread</b>. That is the single thread associated with the Channel for
+ * a Filter Instance. All invocations of that instance's methods are made by the Filter Dispatch Thread.
+ * <p>
+ * If all accesses/mutations of Filter Members happen on the Filter Dispatch Thread, then those interactions are
+ * serial and threadsafe. This executor enables Authors to do work in uncontrolled threads and then return to
+ * the Filter Dispatch Thread to take advantage of this convenient threading guarantee.
+ * </p>
+ * <p>
+ * Note that implementations of FilterDispatchExecutor will not honour the shut-down methods {@link #shutdown()}
+ * and {@link #shutdownNow()} and will throw a RuntimeException if those methods are called, as the client is
+ * not allowed to shut down the Filter Dispatch Thread.
+ * </p>
+ * @see io.kroxylicious.proxy.filter
+ */
+public interface FilterDispatchExecutor extends ScheduledExecutorService {
```

Comment:
extending ScheduledExecutorService is nice in that other technology may be written against ScheduledExecutorService, ExecutorService or Executor. So this can be passed through. We have already seen in the OAuthBearer Filter that it's nice to have a scheduled pool available to delay work.

On the other hand we expose methods like `shutdown` and `shutdownNow` to the client, when we _really_ do not want a Filter to be able to shut down it's eventloop.

Any thoughts on this?

Reply:
I think we should probably wrap and throw unsupported op for shutdown/shutdownNow.